src/cpp/rtps/writer/StatefulWriter.cpp

```diff
@@ -596,74 +588,115 @@ void StatefulWriter::send_any_unsent_changes()
     else
     {
         RTPSWriterCollector<ReaderProxy*> relevantChanges;
-        StatefulWriterOrganizer notRelevantChanges;
         bool force_piggyback_hb = false; // Force piggyback HB if old samples not acknowledged.
 
         NetworkFactory& network = mp_RTPSParticipant->network_factory();
         locator_selector_.reset(true);
         network.select_locators(locator_selector_);
         compute_selected_guids();
 
+        RTPSMessageGroup group(mp_RTPSParticipant, this, *this);
```

Comment:
Maybe it is better to send the GAP always after sending a HEARTBEAT, because you can be telling the reader about a CacheChange that reader doesn't know the writer has in the history because doesn't receive the information in the HEARTBEAT.

Reply:
If we are inside `send_any_unsent_changes` it is likely that a heartbeat has been previously sent, either on matched_reader_add or with a periodic one

Reply:
Done